### ISSUE TITLE
Keyboard handler does not propagate escape key events

### ISSUE THREAD

**Issue by bertspaan** [Leaflet/Leaflet#5234]
When using Leaflet with [keyboard navigation enabled](http://leafletjs.com/reference-1.0.2.html#map-keyboard), the propagation of escape keydown events is stopped by [`L.Map.Keyboard`](https://github.com/Leaflet/Leaflet/blob/63fd4edc76893ab2a2f83d54e703e0a4da73de7b/src/map/handler/Map.Keyboard.js#L159-L166):

```js
_onKeyDown: function (e) {
  var key = e.keyCode;
  if (...) {
    // ...
  } else if (key === 27) { // 27 is escape key's keyCode
    map.closePopup();
  } else {
    return;
  }
  L.DomEvent.stop(e);
}
```

I need to capture escape key events in my application, but creating a Leaflet [`keypress`](http://leafletjs.com/reference-1.0.2.html#map-keypress) handler with `map.on` — or using `map._container.addEventListener` directly — does not work.

Is this intentional? Why was the check for [`map._popup`](https://github.com/Leaflet/Leaflet/commit/1eae1719bb51cef00382b882ec368c5238c873e7) removed in 1eae1719bb51cef00382b882ec368c5238c873e7?

**Comment by perliedman** [comment]
Looking at the history for #3000, which 1eae1719bb51cef00382b882ec368c5238c873e7 is part of, it's probably just a refactor: the conditional isn't needed since it's ok to call `closePopup` even if there isn't already a popup open. Unfortunately, it also causes the issue you're seeing, since the escape key event will always be stopped and never propagate.

@snkashis do you have an opinion here? I guess reintroducing the check for an open popup would be good after all?

**Comment by snkashis** [comment]
Yes, reintroducing check sounds fine to me.

**Comment by droganov** [comment]
Escape key does not emit `keypress` event

http://jsbin.com/qohomibube/edit?js,console,output

please add support for `keydown` and `keyup`

**Comment by IvanSanchez** [comment]
I'm trying to create a unit test for this, but it seems to be quite complicated, because the [keyboard events spec is a bit weird](https://w3c.github.io/uievents/#events-composition-canceling):

> If a `keydown` event is canceled then any Composition Events that would have fired as a result of that `keydown` SHOULD not be dispatched

No composition events means no "character value" is generated, which means no `beforeinput` event, which means no legacy `keypress` event.

The problem here is that utilities for firing synthetic keyboard events such as [`happen`](https://github.com/tmcw/happen) (or `prosthetic-hand` it it were to handle keyboard, [or others](https://github.com/tmcw/happen#see-also)) are simplistic in this regard - they aim to overcome browser differences, not to implement cancellation behaviour.

TL;DR: I've got a PR, but it's impossible to make a unit test for it.

**Comment by pshassett** [comment]
The esc key still doesn't seem to raise a keypress event in chrome 75.0 or firefox 68.0.
```js
map.on('keypress', function(e){
var event = e.originalEvent
alert(event.key)
});
```
Am I missing something in my implementation or could this a browser-based problem?

**Comment by IvanSanchez** [comment]
You're missing the deprecation, see https://developer.mozilla.org/en-US/docs/Web/API/Element/keypress_event

If the "close popup on pressing escape" behaviour is broken, please open a new bug instead.

**Comment by pshassett** [comment]
Got it, thanks for the quick response!

**Comment by Deepak7985** [comment]
I have a pop up form on which i need to submit some date ...but when i pressed submit the listener of the popup is not triggered ..instead the backgound page's listener is triggered .How can i solve this issue.
